prog/types.go

```diff
@@ -694,16 +696,24 @@ func foreachTypeImpl(meta *Syscall, preorder bool, f func(t Type, ctx TypeCtx))
 				break // prune recursion via pointers to structs/unions
 			}
 			seen[a] = true
-			for i := range a.Fields {
-				rec(&a.Fields[i].Type, dir)
+			for i, f := range a.Fields {
```

Comment:
I see that PtrType.ElemDir is used in more places:
```
syzkaller$ find -name "*.go" -exec grep "ElemDir" {} \; -print
			ElemDir:    genDir(args[0]),
./pkg/compiler/types.go
	inner, calls := r.generateArg(s, a.Elem, a.ElemDir)
./prog/rand.go
		elem, elemDir = t1.Elem, t1.ElemDir
			typ, elem, elemDir = anyPtr, anyPtr.Elem, anyPtr.ElemDir
./prog/encoding.go
			if err := ctx.validateArg(arg.Res, typ.Elem, typ.ElemDir); err != nil {
./prog/validation.go
	ElemDir1 Dir
	return fmt.Sprintf("ptr[%v, %v]", t.ElemDir, t.Elem.String())
	return MakePointerArg(t, dir, 0, t.Elem.DefaultArg(t.ElemDir))
			rec(&a.Elem, a.ElemDir)
./prog/types.go
		res := syzType.Elem.DefaultArg(syzType.ElemDir)
		res := ctx.genArg(syzType.Elem, syzType.ElemDir, a)
./tools/syz-trace2syz/proggen/proggen.go
```

Likely field direction needs to be used in these places as well.

Then it may be handy to have a helper function along the lines of:
```
func (f *Field) dir(default Dir) Dir {
  if f.HasDir {
    return f.Dir
  }
  return default
}
```